config: add new `docstring-code-format` knob

BurntSushi opened the pull request:
This PR does the plumbing to make a new formatting option, `docstring-code-format`, available in the configuration for end users. It is disabled by default (opt-in). It is opt-in at least initially to reflect a conservative posture. The intent is to make it opt-out at some point in the future.

This was split out from #8811 in order to make #8811 easier to merge. Namely, once this is merged, docstring code snippet formatting will become available to end users. (See comments below for how we arrived at the name.)

Closes #7146

## Test Plan

Other than the standard test suite, I ran the formatter over the CPython and polars projects to ensure both that the result looked sensible and that tests still passed. At time of writing, one issue that currently appears is that reformatting code snippets trips the long line lint: https://github.com/BurntSushi/polars/actions/runs/7006619426/job/19058868021

BurntSushi commented:
There's also a question of whether this ought to be a boolean option at all. @konstin raised this here: https://github.com/astral-sh/ruff/pull/8811#discussion_r1403255043

That is, we might want the formatter to specifically raise an error if it comes across a snippet that is invalid Python. I somewhat think this would be better as a lint, but it would likely be much easier to implement in the formatter. If it were a lint, then we'd need to refactor the code snippet logic out of the formatter into something more generalized.

MichaReiser reviewed (approved):
The change itself looks good to me but I would prefer changing the option name (also on `PyFormatOptions`). 

Rustfmt uses [`format_code_in_doc_comments`](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#format_code_in_doc_comments) which we could change to `format_code_in_docstrings`. 

I prefer the name because

* It's unclear to me what `snippets` means without prefixing it with `code-` which I find to verbose
* `format-embedded` might be ambiguous if we start formatting embedded language, let's say SQL or CSS in your Python code (the other way round)
* I'm okay with `format-docstring-code`, although I find rustfmt's naming clearer.

I'm okay with using a boolean. I don't think we should bail formatting just because of an incorrect example. Ideally we would emit a warning, but that's tracked as a separate issue. 

The other question: Should this be gated behind preview mode (sorry if this is part of your other PR)? If so, should we emit a warning if it is enabled but the formatter is run without preview mode?

zanieb commented:
I like `format-code-in-docstrings`. I prefer clarity in settings, a little verbosity goes a long way. As I suggested before, I'm also into `format-docstring-code`. Another option is `format-docstring-code-snippets`. I think we can consider non-boolean options in the future, but we should start with the simple case.

> Should this be gated behind preview mode (sorry if this is part of your other PR)? If so, should we emit a warning if it is enabled but the formatter is run without preview mode?

The idea is opt-in without preview opt-out (i.e. on by default) with preview. Having a "double" opt-in seems overkill.

konstin reviewed (approved):
`format-code-in-docstrings` sounds good

BurntSushi commented:
We have a winner. `format-code-in-docstrings` it is!

charliermarsh commented:
Works for me :)

BurntSushi commented:
I think this was talked about internally and not here on the PR, but we ended up settling on `docstring-code-format` for the name instead. I think we all originally liked `format-code-in-docstrings`, but once we realized we were going to add another option for controlling the line width in docstring code snippets, the naming became a little trickier. With `docstring-code-format`, there's a natural name for line width: `docstring-code-line-width`. That is, the `docstring-code-` prefix acts as a namespace of sorts with the noun first and the verb second. But `format-code-in-docstrings` doesn't really have that property, with the verb first and the noun coming afterwards.

MichaReiser commented:
Note: `docstring-code-line-width` should be named `docstring-code-line-length` to align with the global `line-length` setting. We only use `line-width` internally (and we may want to perform an internal rename to use consistent naming internally and externally)

BurntSushi commented:
All righty, I think this PR is now in good spot where another review would be good. The probable changes since the last time you looked:

* The `docstring-code-line-length` option has been added. **It defaults to `dynamic`**, but users can also set a fixed integer too.
* I've added an integration-level test to sanity check that both docstring code options are actually working.
* I've polished up the docs at the setting level and made some small edits to the existing prose formatter docs. I stopped short of adding a new section because I kind of felt like the docs on the setting itself might cover things, but happy to write more if folks think it would be useful.

charliermarsh commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -2896,6 +2896,154 @@ pub struct FormatOptions {
+
+    /// Whether to format code snippets in docstrings.
+    ///
+    /// When this is enabled, Python code in the format of doctests
```
Nit: I'd remove "in the format of doctests", and just say, "Python code within docstrings", since this also applies to reStructuredText and Markdown.

charliermarsh commented on crates/ruff_workspace/src/options.rs:
````diff
@@ -2896,6 +2896,154 @@ pub struct FormatOptions {
+    ///     pass
+    /// ```
+    ///
+    /// will be reformatted (assuming the rest of the options are set
````
I randomly tend to add ellipses to the start of a sentence like this, but it's totally personal preference :)

BurntSushi replied:
Yeah I think I do that sometimes. I'll add it here.

charliermarsh commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -2896,6 +2896,154 @@ pub struct FormatOptions {
+    /// formatter would otherwise write invalid Python code, then the code
+    /// example is ignored by the formatter and kept as-is.
+    ///
+    /// Currently, doctest, Markdown, reStructuredText literal blocks and
```
Nit: Oxford comma for consistency :joy: (after "literal blocks")

BurntSushi replied:
Oof. That's a knife to the heart. In the name of consistency we do evil things.

charliermarsh commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -2896,6 +2896,154 @@ pub struct FormatOptions {
+    /// line length limit.
+    ///
+    /// For example, when this is set to `20` and `docstring-code-format` is
+    /// enabled, then this code:
```
Excellent.

charliermarsh commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -2896,6 +2896,154 @@ pub struct FormatOptions {
+    /// This only has an effect when the `docstring-code-format` setting is
+    /// enabled.
+    ///
+    /// The default value for this setting is `"dynamic"`. This setting has the
```
I would say "This value has the effect" rather than "This setting has the effect", since the setting is `docstring-code-line-length`. Or, like:

```
The default value for this setting is `"dynamic"`, which has the effect of ensuring...
```

charliermarsh reviewed (commented):
This looks good, but can I request that we add a section like:


```markdown
## Docstring formatting

The Ruff Formatter is capable of formatting Python code within docstrings...
```

...to `formatter.md`, just after its `## Configuration` block?

BurntSushi commented on docs/configuration.md:
```diff
@@ -71,6 +71,20 @@ If left unspecified, Ruff's default configuration is equivalent to:
+    #
+    # This only has an effect when the `docstring-code-format` setting is
+    # enabled.
+    docstring-code-line-length = "dynamic"
```
Is it worth keeping them here, but changing `docstring-code-format` to its current default value of `false`?

charliermarsh replied:
Yeah, I think that's fine for visibility. You can mention that it's opt-in for now.

BurntSushi commented:
@charliermarsh I've added prose docs and fixed up your comments. A review of the prose docs would be great! Thanks.

charliermarsh reviewed (approved):
Excellent, thanks!

MichaReiser commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -2896,6 +2896,153 @@ pub struct FormatOptions {
+
+    /// Whether to format code snippets in docstrings.
+    ///
+    /// When this is enabled, Python code examples within docstrings is
```
```suggestion
    /// When this is enabled, Python code examples within docstrings are
```

MichaReiser commented on crates/ruff_workspace/src/options.rs:
````diff
@@ -2896,6 +2896,153 @@ pub struct FormatOptions {
+    ///     pass
+    /// ```
+    ///
+    /// If a code snippt in a docstring contains invalid Python code or if the
+    /// formatter would otherwise write invalid Python code, then the code
+    /// example is ignored by the formatter and kept as-is.
````
Do I remember correctly that we assume a code snipped uses Python when using Markdown or restructured text without specifying the language? If so, should we mention it in the documentation? While we aren't erroring or warning about invalid code snippets just yet, we may want to do so in the future.

charliermarsh replied:
I believe it's mentioned here: https://github.com/astral-sh/ruff/pull/8854/files#diff-d5764d76a6dcbb6934375ce2aec4a33aff239beab6ea71c447e624f9c973b63dR151

BurntSushi replied:
I've also added a mention of that here in the setting docs.

cpcloud commented:
This feature is most excellent, thank you for working on it.
